text/0064-refer-to-types-in-empty-namespace.md

```diff
@@ -0,0 +1,159 @@
+`__cedar::String` refers to the builtin primitive type `String`.
+* to "names" like `::A::B` -- that syntax remains an error. `A::B` is already an
+unambiguous, fully-qualified name, and will never have any namespace implicitly
+prepended.
+
+### Change 2: Implicitly falling back on the empty namespace
+
+When the schema refers to an unqualified typename (e.g, `id`) which doesn't
+exist in the active namespace, we should fall back to looking for it in the
+empty namespace.
+Only if it exists in neither the active nor empty namespace should this be an
+error.
+
+E.g., in the example at the top of this RFC, consider the type of the `id`
+attribute of `Demo::User`.
+Since there is no `id` type defined in namespace `Demo`, this is an error today.
+With this change, when we see there is no `Demo::id`, we implicitly look for
+`id` in the empty namespace, and resolving `id` to be the common type `::id`
+defined to be `String` in the empty namespace.
+The user could, of course, write `::id` directly, but with this change, they
+do not have to, in the common case where there is no collision (where `id` is
+not actually defined in both the `Demo` and empty namespaces).
+
+This applies to both entity and common types defined in the empty namespace.
+
+This applies to both the human (RFC 24) and JSON schema syntaxes.
+
+This is reminiscent of, but not dependent on, the existing similar behavior for
+primtiive and extension types, where we first check for an entity or common type
+`String` in the active namespace before falling back on resolving to
+`__cedar::String`.
+
+## Drawbacks
+
+* The schema implementation becomes more complex, as we need more complex
+methods for typename resolution, and also slightly more complexity in the
+schema parser, as `::id` wasn't previously legal syntax (`::` previously was
+only legal between two identifiers).
+
+This is not a breaking change; all existing schemas remain valid with the same
+semantics as they had before this RFC.
+
+## Alternatives
+
+Instead of `::id`, we could use many other syntaxes.
+One proposal in the original
+[cedar#579](https://github.com/cedar-policy/cedar/issues/579) thread was to use
+something in the `__cedar` namespace, e.g., `__cedar::schema::top_level::id`.
+Other possibilities include other special characters, like `^id` or `#id`, or even
+a special keyword, like `root id`, which should be unambiguous in both the human
+and JSON schema grammars.
+
+## Unresolved questions
+
+* Should we also support the `::id` syntax in policies?
+It's totally unnecessary from an expressiveness perspective, as policies do not
+have a notion of active namespace, so the unqualified name `id` always refers to
+the `id` in the empty namespace and is never implicitly prefixed with any other
+namespace.
+But, users who are used to writing `::User` in schemas (for some `User` type
+they defined in the empty namespace), might expect to also refer to that type as
+`::User` in policies.
```

Comment:
Starting a thread here because threaded conversation (attached to a particular point in the text) is easier to follow than the free-flowing conversation on the main RFC thread.

@andrewmwells-amazon said:

> I don't think we need to allow `::Foo` in policies since it's easy to suggest `Foo` in a linter.

@hakanson said:

> I also wondered why `::User` only works in schemas vs also in Cedar Policies (e.g. `principal is ::User`)

Reply:
The arguments against allowing `::User` in policies appear to be
1. it's unnecessary since `principal is User` is already completely unambiguous -- this is argued in the RFC text quoted above
2. we can catch this and provide an informative error message as suggested by @andrewmwells-amazon 

I'll add a third one, which is that we can always go back and support this later, even if we don't right now.

The arguments _for_ appear to be
1. consistency, since users will write `::User` in schemas and might expect to be able to write `::User` in policies as well (as argued in the RFC text and also by @hakanson)

Reply:
I agree with @cdisselkoen this can be added to policies later (if needed), since "best practice" to have a namespace, and this fixes the gap from mixing no namespace with a namespace in the same schema.  Would there be plans to add a "shadows warning" when validating the schema like with builtin Cedar names?

> The name `String` shadows a builtin Cedar name. You'll have to refer to the builtin as `__cedar::String`.

> The name `User` shadows a un-namespaced name. You'll have to refer to the  un-namespaced as `::User`.

Reply:
> The name User shadows a un-namespaced name. You'll have to refer to the un-namespaced as ::User
Because policies aren't namespaced, this doesn't come up. `User` can only refer to `User` in the empty namespace.

Reply:
Or maybe I misunderstand you and you're asking if we'll get a warning in the schema when in namespace `Bar` I define and mention `User` and there is also a type `::User`. In that case, yes I think we could give a warning, but we won't want it to error to preserve backwards compatibility.

Reply:
@andrewmwells-amazon - in the schema (not policy) as a schema validation warning similar to the other shadows errors

Reply:
We should definitely put in warnings like that, during validation

Reply:
Based on this thread, I lean towards not allowing `::EntityType` in policies, but making sure we add a special error message / help text in case a user (reasonably) tries to write this.

Reply:
Updated the "Unresolved questions" section of the RFC with the conclusion of this thread.  Also added a "Change 3" to this RFC specifying the validation warning (not the policy-parsing warning) discussed in this thread.